File: longest_valid_parentheses_32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        max_length = 0

        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    max_length = max(max_length, i - stack[-1])

        return max_length


s = Solution()
s.longestValidParentheses("(()")
# s.longestValidParentheses("()(()")

File: longest_valid_parentheses_32.js
/**
 * @param {string} s
 * @return {number}
 */

function longestValidParentheses(s) {
  const stack = [-1];
  let maxLength = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") {
      stack.push(i);
    } else {
      stack.pop();
      if (stack.length === 0) {
        stack.push(i);
      } else {
        maxLength = Math.max(maxLength, i - stack.at(-1));
      }
    }
  }
  return maxLength;
}

longestValidParentheses("(()");
longestValidParentheses(")()())");
longestValidParentheses("()(()");
